### ISSUE TITLE
Implementation of other imaging modes for JPK files

### ISSUE THREAD

**Issue by rg314** [AFM-analysis/afmformats#9]
Hi @paulmueller I've been thinking about implementing other imaging modes for the jpk_reader.py. I've written some stuff to parse creep or stress relaxation data from '.txt' exported JPK files just not on the .jpk-force files. Would be good to discuss how to implement it into the current afmformats package.

**Comment by paulmueller** [comment]
Sure, I am open for suggestions. You could probably use the [fmt_ntmdt_txt](https://github.com/AFM-analysis/afmformats/blob/master/afmformats/fmt_ntmdt_txt.py) as a template. It is important that the "detect" function correctly identifies the file, otherwise there will be ambiguities.

I would suggest the name "fmt_jpk_txt". Can you discriminate between force-distance/stress relaxation/creep compliance data just from looking at the .txt file? That would be good.

For "mode" strings, I would suggest "stress-relaxation" and "creep-compliance", if that makes any sense.

**Comment by rg314** [comment]
Okay I'll check the [fmt_ntmdt_txt](https://github.com/AFM-analysis/afmformats/blob/master/afmformats/fmt_ntmdt_txt.py) file. It's easy to discriminate between f-d and stress relaxation/creep because there are three segments in the latter case. However, it's harder to discriminate between stress relaxation and creep but I will have a think about this.

Yes, the mode naming makes sense. My concern is more about the parsing of the metadata and what format it needs to be in. I'm also keen to look into identifing creep and stress relaxation directly from the jpk-force file as this saves having to output .txt files from the jpk software.

**Comment by paulmueller** [comment]
@rg314 Do you have exemplary data files for creep compliance and stress relaxation that you can share publicly (CC0)?

**Comment by rg314** [comment]
Hi @paulmueller, this issue completely passed my mind. I've got some creep files that I can share publicly. It's a little hand wavy but here's a code snipped for reading them in. 

```
    def _jpk_text_loader(self):
        
        with open(self.file) as f:
            metadata_list_string = [x.strip("# ").strip("\n").split(": ") for x in f.readlines() if "#" in x]

        segment = "instrument"
        metadata = defaultdict(dict)
        for v in metadata_list_string:
            if len(v) > 1:
                k = v[0]
                try:
                    v = literal_eval(v[-1])
                except:
                    v = v[-1]
                if k == "segment":
                    segment = v
                    metadata[segment][k] = v
                else:
                    metadata[segment][k] = v

        self.metadata = metadata

        if "extend" in self.metadata.keys():
            self.experiment_type = 'force_displacement'

        if "pause" in self.metadata.keys():
            self.experiment_type = 'creep'

        with open(self.file) as f:
            data = re.findall(r"^[\d.E\s-]+$", f.read().strip(), re.MULTILINE)

        self.data = dict()
        i = 0
        for metadata_segment in self.metadata.keys():
            data_name = self.metadata[metadata_segment].get("segment", None)
            if data_name is not None:
                self.data[data_name] = pd.read_csv(StringIO(data[i]), delimiter=' ', names=self.metadata[metadata_segment]['columns'].split())
                i += 1
        
        self.springConstant = eval(self.metadata["extend"]["springConstant"])
```

**Comment by paulmueller** [comment]
I'm thinking it might be better to directly load the data from the .jpk-force files. Do you also have such files lying around?

**Comment by rg314** [comment]
Add 2x .txt and 2x .jpk-force to the PR https://github.com/AFM-analysis/afmformats/pull/14

**Comment by paulmueller** [comment]
I've managed to implement support for creep-compliance data in .jpk-force files. I will add your files as test files to the repository.

Is there a particular reason why you added **two** datasets (e.g. created from different JPK software versions or different setups)?

And do you happen to have stress-relaxation measurements lying around as well?

**Comment by rg314** [comment]
Yes, the measurements were made at different times, on different instruments, and on different materials (brain + polymer gel). I thought it would be good to include different cases for testing.

Nope, I don't have any stress relaxation lying around but I could make sure measurements next week, would this be useful @paulmueller?

**Comment by paulmueller** [comment]
OK, then I will include both files. Which one is the brain and which one the polymer-gel?
Yes, stress-relaxation measurements will definitely be useful. I's always good to have some reference measurements for testing. Thanks!
